sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/OrderedListUserState.java

```diff
@@ -0,0 +1,240 @@
+          .getStateKeyBuilder()
+          .getOrderedListUserStateBuilder()
+          .getRangeBuilder()
+          .setStart(minTimestamp.getMillis())
+          .setEnd(limitTimestamp.getMillis());
+
+      // TODO: consider use cache here
+      CachingStateIterable<TimestampedValue<T>> persistentValues =
+          StateFetchingIterators.readAllAndDecodeStartingFrom(
+              Caches.noop(),
+              this.beamFnStateClient,
+              getRequestBuilder.build(),
+              this.timestampedValueCoder);
+
+      // Make a snapshot of the current pendingRemoves and use them to filter persistent values.
+      // The values of pendingRemoves are kept, so that they will still be accessible in
+      // pre-existing iterables even after a sort key is removed.
+      TreeRangeSet<Instant> pendingRemovesSnapshot = TreeRangeSet.create(pendingRemoves);
+      Iterable<TimestampedValue<T>> persistentValuesAfterRemoval =
+          Iterables.filter(
+              persistentValues, v -> !pendingRemovesSnapshot.contains(v.getTimestamp()));
+
+      return Iterables.mergeSorted(
+          ImmutableList.of(persistentValuesAfterRemoval, valuesInRange),
+          Comparator.comparing(TimestampedValue::getTimestamp));
+    }
+
+    return valuesInRange;
+  }
+
+  public Iterable<TimestampedValue<T>> read() {
+    checkState(
+        !isClosed,
+        "OrderedList user state is no longer usable because it is closed for %s",
+        request.getStateKey());
+
+    return readRange(Instant.ofEpochMilli(Long.MIN_VALUE), Instant.ofEpochMilli(Long.MAX_VALUE));
+  }
+
+  public void clearRange(Instant minTimestamp, Instant limitTimestamp) {
+    checkState(
+        !isClosed,
+        "OrderedList user state is no longer usable because it is closed for %s",
+        request.getStateKey());
+
+    // Remove items (in a collection) in the specific range from pendingAdds.
+    // The old values of the removed sub map are kept, so that they will still be accessible in
+    // pre-existing iterables even after the sort key is cleared.
+    pendingAdds.subMap(minTimestamp, true, limitTimestamp, false).clear();
+    if (!isCleared) {
+      pendingRemoves.add(
+          Range.range(minTimestamp, BoundType.CLOSED, limitTimestamp, BoundType.OPEN));
+    }
+  }
+
+  public void clear() {
+    checkState(
+        !isClosed,
+        "OrderedList user state is no longer usable because it is closed for %s",
+        request.getStateKey());
+    isCleared = true;
+    // Create a new object for pendingRemoves and clear the mappings in pendingAdds.
+    // The entire tree range set of pendingRemoves and the old values in the pendingAdds are kept,
+    // so that they will still be accessible in pre-existing iterables even after the state is
+    // cleared.
+    pendingRemoves = TreeRangeSet.create();
+    pendingAdds.clear();
+  }
+
+  public void asyncClose() throws Exception {
+    isClosed = true;
+
+    if (!pendingRemoves.isEmpty()) {
```

Comment:
Can processing the pendingRemoves and pendingAdds be combined into a single StateRequest across the FnApi? Fewer roundtrips would probably be better.

Reply:
Also, it'd be nice if we could clear multiple ranges in a single request.

Reply:
> Can processing the pendingRemoves and pendingAdds be combined into a single StateRequest across the FnApi? Fewer roundtrips would probably be better.

Hmmmm, again this is about whether we want to be consistent with other states and reuse the existing StateRequest OR create a new one particularly for OrderListState. 

We had this discussion previously at https://github.com/apache/beam/pull/30317#discussion_r1500878413, and I think the decision was to follow the existing states and reuse the current StateRequest, which means we will have Append to do the insertion only and Clear to do the deletion. 

Notice that there is no extra field in Clear request message; it is invoked on a state key that contains a single range. That's why we only do one range at a time.